src/filter/kuadrant_filter.rs

```diff
@@ -1,22 +1,24 @@
 use crate::action_set_index::ActionSetIndex;
-use crate::filter::operations::{
-    GrpcMessageReceiverOperation, GrpcMessageSenderOperation, Operation,
-};
+use crate::configuration::FailureMode;
+use crate::filter::operations::{EventualOperation, Operation};
+use crate::runtime_action::IndexedRequestResult;
 use crate::runtime_action_set::RuntimeActionSet;
-use crate::service::{GrpcErrResponse, GrpcRequest, HeaderResolver, Headers};
+use crate::service::errors::ProcessGrpcMessageError;
+use crate::service::{DirectResponse, GrpcRequest, HeaderResolver, Headers, IndexedGrpcRequest};
```

Comment:
There is quite some additional coupling added here... While I can see why the changes made make this required, I can't judge on the merits of them.